fixtures/tests/_fixtures/test_monkeypatch.py

```diff
@@ -72,14 +83,112 @@ def test_delete_missing_attribute(self):
             fixture.cleanUp()
             self.assertFalse('new_attr' in globals())
 
+    def _check_static_or_class_method(self, name, new_method):
+        oldmethod = getattr(C, name)
+        oldmethod_inst = getattr(C(), name)
+        fixture = MonkeyPatch(
+            'fixtures.tests._fixtures.test_monkeypatch.C.%s' % name,
+            new_method)
+        with fixture:
+            getattr(C, name)()
+            getattr(C(), name)()
+
+        restored_method = getattr(C, name)
+        restored_method_inst = getattr(C(), name)
+        self.assertEqual(oldmethod, restored_method)
+        self.assertEqual(oldmethod, restored_method_inst)
+        self.assertEqual(oldmethod_inst, restored_method)
+        self.assertEqual(oldmethod_inst, restored_method_inst)
+        restored_method()
+        restored_method_inst()
+
     def test_patch_staticmethod(self):
-        oldfoo = C.foo
+        self._check_static_or_class_method('foo_staticmethod', fake_no_args)
+
+    def test_patch_staticmethod_with_staticmethod(self):
+        self._check_static_or_class_method('foo_staticmethod',
+                fake_staticmethod)
```

Comment:
In another comment you mentioned sitting back and deciding on semantics. I completely agree with that and will add my observation of the semantics of mox. Mox works in this case, and actually requires it. test_patch_staticmethod where a staticmethod is being replaced with a non decorated function would fail because it doesn't handle the promotion of a new function to a staticmethod. Of course that doesn't mean this needs to follow that example, but it makes switching from one to another much easier.

I need to think more on the idea of patching the descriptor as-is, because I take your point about the "assert class if F" example above. Rather than pulling the descriptor and using that directly I will look into taking the result of getattr() and rewrapping with static/classmethod and see how that affects things.